src/main/java/bio/terra/common/tracing/TracingHook.java

```diff
@@ -0,0 +1,205 @@
+package bio.terra.common.tracing;
+
+import bio.terra.stairway.*;
+import io.opencensus.common.Scope;
+import io.opencensus.trace.*;
+import io.opencensus.trace.propagation.SpanContextParseException;
+import java.util.Base64;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A {@link StairwayHook} to add support for tracing execution of Stairway flights with OpenCensus.
+ *
+ * <p>This hook uses Spans at 3 different levels. From lowest to highest:
+ * <li>A Span around the execution of each Step, one Span per Step execution. All of the {@link
+ *     Step#doStep(FlightContext)} {@link Step#undoStep(FlightContext)} will be executed in the
+ *     context of this Span.
+ * <li>A Span around the execution of each Flight, one Span per {@link Flight#run()}. Note that a
+ *     single Flight may be run multiple times if its execution is paused and resumed. Each Step
+ *     Span is a child Span of a Flight Span.
+ * <li>A linked Span for the entire Flight submission. Each Flight Span has a parent link to the
+ *     submission span. See <a href="https://opencensus.io/tracing/span/link/">OpenCensus Link</a>.
+ *     The submission span is used to correlate all of the Flight Spans together. The submission
+ *     Span may be created outside of the context of this hook and passed in with the input
+ *     FlightMap; this allows all of the Flight Spans to be linked to the request Span that
+ *     initiated the Flight. See {@link #serializedCurrentContext()} {@link
+ *     #SUBMISSION_SPAN_CONTEXT_MAP_KEY}. If no submission Span is passed into the Flight, a dummy
+ *     Submission span is created by this hook so that all Flight Spans are still correlated.
+ *
+ *     <p>If a Span is not ended, it will not be exported to the tracing service.
+ *
+ * @see <a href="https://opencensus.io/tracing/">https://opencensus.io/tracing/</a>
+ */
+public class TracingHook implements StairwayHook {
+  private final Tracer tracer = Tracing.getTracer();
+  private final Logger logger = LoggerFactory.getLogger(TracingHook.class);
+
+  // Prefixes to use for Span names. Standard prefixes make it easier to search for all Spans of
+  // different types.
+  private static final String SUBMISSION_NAME_PREFIX = "stairway/submission/";
+  private static final String FLIGHT_NAME_PREFIX = "stairway/flight/";
+  private static final String STEP_NAME_PREFIX = "stairway/step/";
+
+  /** Map of Flight IDs to Flight Spans. */
+  // DO NOT SUBMIT: Consider whether we want to have a FlightHook to match StepHook instead of
+  // storing per Flight data in a global map. Note that the FlightHook would have to be able to make
```

Comment:
FlightHook would be a straightforward extension.

Reply:
Any preference for or against adding it?

Reply:
If it would be useful, let's add it right away before we get more hook code written!
It would be small work.